src/test/java/org/dmfs/jems/optional/adapters/SingleOptionalTest.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2018 dmfs GmbH
+ *
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.dmfs.jems.optional.adapters;
+
+import org.dmfs.jems.single.Single;
+import org.dmfs.optional.Absent;
+import org.dmfs.optional.Optional;
+import org.dmfs.optional.Present;
+import org.junit.Test;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.dmfs.jems.hamcrest.matchers.AbsentMatcher.isAbsent;
+import static org.dmfs.jems.hamcrest.matchers.PresentMatcher.isPresent;
+import static org.junit.Assert.assertThat;
+
+
+/**
+ * Test for {@link SingleOptional}.
+ *
+ * @author Gabor Keszthelyi
+ */
+public final class SingleOptionalTest
+{
+
+    @Test
+    public void test()
+    {
+        assertThat(new SingleOptional<>(() -> new Present<>(3)), isPresent(3));
+        assertThat(new SingleOptional<>(Absent::absent), isAbsent());
+    }
+
+
+    @Test
+    public void testThatValueIsCached()
+    {
+        AtomicInteger counter = new AtomicInteger(3);
```

Comment:
> A `Single` represents a tuple with exactly one value.

That means technically a `Single` should always return an (immutable) value equal to the one in the tuple. A `Single` is quite different from a Factory in that regard. Returning `ContentValues` (like I did in the task provider) from a Single is not a good idea which can easily break if `Frozen` is being used, because if anyone changes the instance it will be changed for everyone holding a reference to the same instance.

My point is that creating a `Single` from mutable value is incorrect usage and you're testing the class under false premisses. Also `Frozen` should not be used to guard against unstable values of a `Single`. It's meant to avoid costly revaluation of an expensive value.

Reply:
I've updated the test to check for the number of evaluations of the provided `Single`'s `value()` method instead of changing the value.

As you write here that `Single` is different from `Factory`, shall we create and use a `Factory/Producer` type instead? What should be the name instead of `SingleOptional`?

Or shall we instead try always to avoid this with more composition?

Reply:
After some thinking, I suppose it may be better to not use factories to compose the objects... so I am going to remove this `SingleOptional`, and just implement `Conditional` directly.

Reply:
Done